packages/snaps-controllers/src/snaps/SnapController.ts

```diff
@@ -2250,12 +2260,51 @@ export class SnapController extends BaseController<
     handler: handlerType,
     request,
   }: SnapRpcHookArgs & { snapId: SnapId }): Promise<unknown> {
+    const permissionName = handlerEndowments[handlerType];
+    const hasPermission = this.messagingSystem.call(
+      'PermissionController:hasPermission',
+      snapId,
+      permissionName,
+    );
+
+    if (!hasPermission) {
+      throw new Error(
+        `Snap "${snapId}" is not permitted to use "${permissionName}".`,
+      );
+    }
+
+    if (permissionName === SnapEndowments.Rpc) {
+      const subject = this.messagingSystem.call(
+        'SubjectMetadataController:getSubjectMetadata',
+        origin,
+      );
+      const isSnap = subject?.subjectType === SubjectType.Snap;
```

Comment:
Is it safe to assume everything else is a dapp?

Reply:
It could be dapps, or the extension itself. We could maybe check `subjectType === SubjectType.Website`, but I'm not sure if that would work well with how `SubjectMetadataController` is set up. It caches subjects, but not indefinitely. Snaps are always present AFAIK.